virtual_rainforest/main.py

```diff
@@ -81,6 +85,124 @@ def configure_models(
     return models_cfd
 
 
+def extract_timing_details(
+    config: dict[str, Any]
+) -> tuple[datetime64, datetime64, timedelta64]:
+    """Extract timing details for main loop from the model configuration.
+
+    Args:
+        config: The full virtual rainforest configuration
+
+    Raises:
+        InitialisationError: If the model is set to end before it starts, the units of
+            update interval aren't valid, or if the interval is too small for the model
+            to ever update.
+    """
+
+    # First extract start and end times
+    start_time = datetime64(config["core"]["timing"]["start_date"])
+    end_time = datetime64(config["core"]["timing"]["end_date"])
+
+    # Catch bad time dimensions
+    try:
+        raw_interval = pint.Quantity(config["core"]["timing"]["main_time_step"]).to(
+            "minutes"
+        )
+    except (pint.errors.DimensionalityError, pint.errors.UndefinedUnitError):
+        log_and_raise(
+            "Units for core.timing.main_time_step are not valid time units: %s"
+            % config["core"]["timing"]["main_time_step"],
+            InitialisationError,
+        )
+    else:
+        # Round raw time interval to nearest minute
+        update_interval = timedelta64(int(raw_interval.magnitude), "m")
```

Comment:
This worries me - I don't think we're going to be troubled by it immediately, but why do we need to round? I think this gets a lot easier if we reconsider the grain of the timing loop (see comment below)

Reply:
Oh - comments in the review show up in posting order, not position in file 😖

Reply:
Yeah this is a problem with using `timedelta64`, as far as I can tell you can't supply it with a float and then units. So, e.g. `timedelta64(0.5, "D")` results in a `TypeError`. I've been tackling this by converting units to minutes and then rounding them to minimise precision loss

Reply:
OK - so can only have integer values (but numpy can combine them sanely):

```python
 np.timedelta64(1, 'm') + np.timedelta64(12, 's')
numpy.timedelta64(72,'s')
```

I may be being paranoid here, but we could easily go finer:

```python
In [83]: np.datetime64('2012-06-12 01:00:00') + np.timedelta64(np.timedelta64(32,'m'))
Out[83]: numpy.datetime64('2012-06-12T01:32:00')

In [86]: np.datetime64('2012-06-12 01:00:00') + np.timedelta64(np.timedelta64(32,'m'), 'us')
Out[86]: numpy.datetime64('2012-06-12T01:32:00.000000')
```

Maybe microsecond is going too far 😄 . We can't go finer as the range of representable dates is [too short](https://numpy.org/doc/stable/reference/arrays.datetime.html#datetime-units) 😁

Reply:
Hahahaha, did consider switching to second precision but decided against. Finest scale I can imagine someone realistically wanting is 20 or 30 mins, which does need minute level precision, but sub minute precision is almost certainly never going to be possible

Reply:
Or possible with meaningful output produced within a reasonable simulation time

Reply:
About to start my review, but one thing I want to highlight is that, keeping aside if sub-minute resolution is needed or not - which is for you to decide -, the more steps you do, the higher is going to be the loss of precision due to the numerical accuracy of floating point numbers. Not sure if this might become a problem, but I just flag this here so you can consider it.

Reply:
By the way, `int()` truncates, not rounds float numbers. So `int(0.7) = 0`. I guess this might not be what you want. Maybe `int(round(0.7)) = 1` will be best.